Make Docker Jupyter support to the Version 0.4 as Version 0.2

masquerlin opened the pull request:
@ekzhu 
## Why are these changes needed?

This PR introduces a safer and more controllable execution environment for LLM code execution in version 0.4 by enabling the use of Jupyter inside a container. This enhancement addresses security concerns and provides a more robust execution context. In particular, it allows:

Isolation of code execution via containerized Jupyter environments.

Persistent memory of variables and their values throughout the conversation.

Memory of code execution results to support more advanced reasoning and follow-up tasks.

These improvements help build a more interactive and stateful LLM-agent programming experience, especially for iterative code generation and debugging scenarios.

## Related issue number

Open #6153

## Checks

- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/code_executors/docker_jupyter/_docker_jupyter.py:
```diff
@@ -0,0 +1,255 @@
+from autogen_core import Component
+from ._jupyter_server import JupyterClient, JupyterConnectable, JupyterConnectionInfo
+
+class CodeResult(BaseModel):
```
We should be using the type from core: 

https://github.com/microsoft/autogen/blob/3425d7dc2c5bd53c90162ed0396e340968f0056f/python/packages/autogen-core/src/autogen_core/code_executor/_base.py#L27

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/code_executors/docker_jupyter/_docker_jupyter.py:
```diff
@@ -0,0 +1,255 @@
+
+    output: str
+    
+class IPythonCodeResult(CodeResult):
```
```suggestion
class DockerJupyterCodeResult(CodeResult):
```

ekzhu reviewed (changes requested):
We need to add tests that can run on GitHub CI.

masquerlin commented:
@ekzhu 
Thanks for the feedback! I've made the following updates based on your comments:

1. Added tests under python/packages/autogen-ext/tests/code_executors/test_docker_jupyter_code_executor.py. The tests run successfully on my end (Docker is required).

2. Refactored python/packages/autogen-ext/src/autogen_ext/code_executors/docker_jupyter/_jupyter_server.py to use an async API throughout.

3. Updated DockerJupyterServer to support usage as an async context manager.

4. Added support for configuring workdir and binddir in DockerJupyterServer, which is now allowed to link to the output_dir of DockerJupyterCodeExecutor. This allows user or AI to track file generation.


![image](https://github.com/user-attachments/assets/ffb17af2-0534-45e9-8ff2-581a5545d4d1)

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/code_executors/docker_jupyter/_docker_jupyter.py:
```diff
@@ -0,0 +1,270 @@
+@dataclass
+class DockerJupyterCodeResult(CodeResult):
+    """(Experimental) A code result class for IPython code executor."""
+    output_files: list[Path] | list[str]
```
Just keep `List[Path]` and remove `List[str]`

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/code_executors/docker_jupyter/_docker_jupyter.py:
```diff
@@ -0,0 +1,270 @@
+    jupyter_server: Union[JupyterConnectable, JupyterConnectionInfo]
+    kernel_name: str = "python3"
+    timeout: int = 60
+    output_dir: Union[Path, str] = None
```
This should be `Path | None = None`.

masquerlin commented:
@ekzhu 
Thank you for the detailed feedback!

I've addressed all the suggested changes:

1. Added a unit test to verify that variables persist across code block executions within the same executor context.

2. Updated the executor to use a temporary directory by default.

3. Implemented the start() function and moved the kernel client initialization into it.

4. Other little suggested changes.

Replaced **kwargs with fully specified arguments cancellationToken in execute_code_blocks.

All tests are passing now. Let me know if there's anything else you'd like me to adjust!

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/code_executors/docker_jupyter/_jupyter_server.py:
```diff
@@ -1,5 +1,5 @@
 from __future__ import annotations
-import docker
+import docker # type: ignore
```
Let's not ignore these imports. If we need to add dependencies we should update `pyproject.toml` in `autogen-ext`.

ekzhu replied:
Add a new entry `docker-jupyter-executor` after this

https://github.com/microsoft/autogen/blob/1035065c4e1f0bd913a544032cc381c55004ffa6/python/packages/autogen-ext/pyproject.toml#L76-L81

Then, update the uv.lock file by running `uv sync --all-extras` in the `python` directory. Make sure your local uv version is set to 0.5.18:

```
uv self update 0.5.18
```

masquerlin commented:
> Thanks @masquerlin I added a few fixes here and there.

Glad to contributed and added this feature. Appreciate the improvements you made.
